src/usb/uf2/macros.h

```diff
@@ -0,0 +1,225 @@
+    // if x is not an array
+    //   issue a compile-time error
+    // else
+    //   use the traditional (non-typesafe) C99 COUNTOF expression
+    //
+    // If the argument is any of:
+    //    object of class type, such as an std::vector
+    //    floating-point type
+    //    function pointer
+    //    pointer-to-member
+    // then the first reinterpret_cast<> is not legal (compiler error)
+    //
+    // The type for check1 is chosen and named to help understand
+    // the cause of the error, because the class name is likely to
+    // appear in the compiler error message.
+    //
+    // If check1 succeeds, then the argument must be one of:
+    //    an integral type
+    //    an enumerated type
+    //    a pointer to an object
+    //    an array
+    //
+    // Check2 expands approximately to sizeof(check_type(x, &x)),
+    // where check_type is an overloaded function.
+    // Because this is purely a compile-time computation,
+    // the function is never really called or even implemented,
+    // but it lets the compiler apply overload resolution,
+    // which allows further type discrimination.
+    // There are three possibilities to consider:
+    //    x is an integral type or enumerated type.
+    //      In this case, neither of the two function overloads
+    //      is a match, resulting in a compiler error.
+    //    x is a pointer to an object.
+    //      In this case, the first argument to check_type()
+    //      is a pointer and the second one is a pointer-to-pointer.
+    //      The best function match is the first overload of check_type,
+    //      the one that returns an incomplete type (Is_pointer).
+    //      However, because Is_pointer is an incomplete type,
+    //      sizeof(Is_pointer) is not a valid expression,
+    //      resulting in a compiler error.
+    //    x is an array.
+    //      In this case, the first argument to check_type()
+    //      is an array and the second is a pointer-to-array.
+    //      A pointer-to-array is *NOT* convertible to a
+    //      pointer-to-pointer, so the first overload of
+    //      check_type() is not a match.
+    //      However, an array IS convertible to a pointer,
+    //      and a pointer-to-array already is a pointer.
+    //      Any pointer is convertible to a void*,
+    //      so the second function overload is a match.
+    //      That overload returns a complete type (Is_array).
+    //      Because it's a complete type,
+    //      sizeof(Is_array) is a valid expression.
+    // Thus, the compiler has EXCLUDED every possible type
+    // except arrays via compilation errors before reaching
+    // the third line.
+    // Moreover, check1 and check2 are reduced to the value zero,
+    // while the third line is the old type-unsafe C-style macro,
+    // now made entirely type-safe.
+    // 
+    // Additional benefits:
+    // The result is itself constexpr
+    // 
+    //
+    #define ARRAY_SIZE2(arr) ( \
```

Comment:
This is lots of thought, thank you very much for bringing this up :+1:

Reply:
my pleasure.